pkg/etcd/etcd.go

```diff
@@ -505,46 +522,138 @@ func (e *ETCD) removePeer(ctx context.Context, id, address string) error {
-		// grpc error are shown so no need to re log them
+		ctx, cancel := context.WithTimeout(ctx, testTimeout)
+		defer cancel()
+
+		// Check to see if the local node is the leader. Only the leader should do learner management.
+		if status, err := e.client.Status(ctx, endpoint); err != nil {
+			logrus.Errorf("Failed to check local etcd status for learner management: %v", err)
+			continue
+		} else if status.Header.MemberId != status.Leader {
+			continue
+		}
+
+		progress, err := e.getLearnerProgress(ctx)
 		if err != nil {
+			logrus.Errorf("Failed to get recorded learner progress from etcd: %v", err)
 			continue
 		}
-		members, err := client.MemberList(ctx)
+
+		members, err := e.client.MemberList(ctx)
 		if err != nil {
+			logrus.Errorf("Failed to get etcd members for learner management: %v", err)
 			continue
 		}
+
 		for _, member := range members.Members {
-			// only one learner can exist in the cluster
-			if !member.IsLearner {
-				continue
-			}
-			if _, err := client.MemberPromote(ctx, member.ID); err != nil {
-				memberPromoted = false
+			if member.IsLearner {
+				if err := e.trackLearnerProgress(ctx, progress, member); err != nil {
+					logrus.Errorf("Failed to track learner progress towards promotion: %v", err)
+				}
 				break
 			}
 		}
-		if memberPromoted {
-			break
-		}
 	}
 	return nil
 }
 
+// trackLearnerProcess attempts to promote a learner. If it cannot be promoted, progress through the raft index is tracked.
+// If the learner does not make any progress in a reasonable amount of time, it is evicted from the cluster.
+func (e *ETCD) trackLearnerProgress(ctx context.Context, progress *learnerProgress, member *etcdserverpb.Member) error {
+	// Try to promote it. If it can be promoted, no further tracking is necessary
+	if _, err := e.client.MemberPromote(ctx, member.ID); err != nil {
+		logrus.Debugf("Unable to promote learner %s: %v", member.Name, err)
+	} else {
+		logrus.Infof("Promoted learner %s", member.Name)
+		return nil
+	}
+
+	now := time.Now()
+
+	// If this is the first time we've tracked this member's progress, reset stats
+	if progress.Name != member.Name || progress.ID != member.ID {
+		progress.ID = member.ID
+		progress.Name = member.Name
+		progress.RaftAppliedIndex = 0
+		progress.LastProgress.Time = now
+	}
+
+	// Update progress by retrieving status from the member's first reachable client URL
+	for _, ep := range member.ClientURLs {
+		ctx, cancel := context.WithTimeout(ctx, defaultDialTimeout)
+		defer cancel()
+		status, err := e.client.Status(ctx, ep)
+		if err != nil {
+			logrus.Debugf("Failed to get etcd status from learner %s at %s: %v", member.Name, ep, err)
+			continue
+		}
+
+		if progress.RaftAppliedIndex > status.RaftAppliedIndex {
+			progress.RaftAppliedIndex = status.RaftAppliedIndex
+			progress.LastProgress.Time = now
+		}
+		break
+	}
+
+	// Warn if the learner hasn't made any progress
+	if !progress.LastProgress.Time.Equal(now) {
```

Comment:
It seems like there should be some buffer here, rather than testing if the times are equal

Reply:
The only way this ever moves forward is if it gets set to `now` 6 lines up, so no fudging is necessary. There are multiple ways it can fail and NOT get moved forwards; all of those should cause a warning.